core/camel-support/src/main/java/org/apache/camel/support/EnumArray.java

```diff
@@ -0,0 +1,174 @@
+
+package org.apache.camel.support;
+
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.function.Function;
+
+/**
+ * This class provides an array-based structure used to store payloads associated with enums. This is
+ * used in the hot path of the core code to allow access to those payloads with constant time and low
+ * memory overhead.
+ *
+ * This data-structure is meant for internal usage of the Camel Core and is not meant for users.
+ * @param <T>
+ */
+public final class EnumArray<T extends Enum<?>> {
+    private final Object[] internalArray;
+    private final T[] values;
+
+    /**
+     * Creates a new EnumArray with a fixed size determined by the length of values
+     * @param values the Enum values (as in Enum.values())
+     */
+    public EnumArray(T[] values) {
+        this.internalArray = new Object[values.length];
+        this.values = values;
+    }
+
+    /**
+     * Whether this arrau contains a payload (value) for the given entry
+     * @param entry the entry to check for
+     * @return true if there is a payload or false otherwise
+     */
+    public boolean contains(T entry) {
+        Object payload = internalArray[entry.ordinal()];
+
+        return payload != null;
+    }
+
+    /**
+     * Gets the payload for the given entry
+     * @param entry the entry to get the payload (must not be null)
+     * @return the payload or false otherwise
+     */
+    public Object get(T entry) {
+        Objects.requireNonNull(entry, "The entry for a enum array cannot be null");
+
+        return internalArray[entry.ordinal()];
+    }
+
+    /**
+     * Gets the payload for the given entry
+     * @param entry the entry to get the payload (must not be null)
+     * @param clazz the class to cast the payload to
+     * @return the payload or false otherwise
+     * @param <K> the payload type
+     */
+    public <K> K get(T entry, Class<K> clazz) {
+        Object tmp = get(entry);
+
+        if (tmp != null) {
+            return clazz.cast(tmp);
+        }
+
+        return null;
+    }
+
+    /**
+     * Computes a new value for the entry if it's present on this instance
+     * @param entry the entry to compute the value for
+     * @param mappingFunction the mapping function that will provide the new value. It takes as a parameter the old value.
+     * @return the new value that was computed for this entry
+     */
+    public Object computeIfPresent(T entry, Function<Object, Object> mappingFunction) {
+        Object payload = get(entry);
+
+        if (payload != null) {
+            payload = mappingFunction.apply(payload);
+
+            set(entry, payload);
+        }
+
+        return payload;
+    }
+
+    /**
+     * Sets the payload for the given entry
+     * @param entry the entry to set the payload
+     * @param payload the payload
+     */
+    public void set(T entry, Object payload) {
+        internalArray[entry.ordinal()] = payload;
```

Comment:
```suggestion
        internalArray[Objects.requireNonNull(entry, "The entry for a enum array cannot be null").ordinal()] = payload;
```
To be homogenous, we should check if it is null here too